### PR TITLE
[Concepts] Add Decls from the outer scope of the current lambda for conversion function constraints

### PR DESCRIPTION
**Pull request by zyn0217**
This fixes the case shown by https://github.com/llvm/llvm-project/issues/64808#issuecomment-1929131611.

Similar to https://github.com/llvm/llvm-project/commit/f9caa12328b265b77221fe7a310d4504673d814a, we have some calls to constraint checking for a lambda's conversion function while determining the conversion sequence.

This patch addresses the problem where the requires-expression within such a lambda references to a Decl outside of the lambda by adding these Decls to the current instantiation scope.

I'm abusing the flag `ForOverloadResolution` of CheckFunctionConstraints, which is actually meant to consider the Decls from parent DeclContexts.

### CODE REVIEW

File: clang/lib/Sema/SemaOverload.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaOverload.cpp:7979]: This isn't really specific to lambdas right?  Doing this on a function seems like it would do the wrong thing, right?

Comment (zyn0217) [reply clang/lib/Sema/SemaOverload.cpp:7979]: Hmm... Yes, the flag would tell the LSI to look into its parents as well, although we don't have any coverage test cases for it yet. Maybe I think we can narrow it down to the inside of CheckFunctionConstraints.

**Comment by zyn0217** [comment]
@erichkeane : I've updated the patch to limit the scope to lambda-only, which is now more appropriate to me. PTAL, thanks!

**Review by erichkeane — COMMENTED** [review]
I think this looks good, but I want @cor3ntin to do the final approval, he touched it last :)

**Comment by zyn0217** [comment]
@cor3ntin WDYT?

File: clang/docs/ReleaseNotes.rst
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:279]:
```suggestion
  expression references to an entity declared outside of the lambda. Fixes
```

File: clang/docs/ReleaseNotes.rst
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:280]:
```suggestion
  (#GH64808)
```

File: clang/lib/Sema/SemaConcept.cpp
Comment (cor3ntin) [inline clang/lib/Sema/SemaConcept.cpp:695]:
```suggestion
  // considered. Turning on the 'ForOverloadResolution' flag results in the
```

**Review by cor3ntin — APPROVED** [review]
LGTM modulo nitpicks

**Pull request merged**
